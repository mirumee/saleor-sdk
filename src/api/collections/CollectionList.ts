import BaseList, { GetPageInfo, MapQueryData } from "../../helpers/BaseList";
import { CollectionList as CollectionListQuery } from "../../queries/gqlTypes/CollectionList";
import { BaseCollection } from "../../fragments/gqlTypes/BaseCollection";

export class CollectionList extends BaseList<
  CollectionListQuery,
  BaseCollection
> {
  getPageInfo: GetPageInfo<CollectionListQuery> = result =>
    result.data.collections?.pageInfo!;

  mapQueryData: MapQueryData<CollectionListQuery, BaseCollection> = data =>
    data.collections?.edges.map(({ node }) => node);
}
