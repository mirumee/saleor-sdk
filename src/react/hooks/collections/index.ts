export { default as useCollectionList } from "./useCollectionList";
