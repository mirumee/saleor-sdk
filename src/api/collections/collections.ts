import ApolloClient from "apollo-client";
import {
  CollectionList as CollectionListQuery,
  CollectionListVariables,
} from "../../queries/gqlTypes/CollectionList";
import { BaseCollection } from "../../fragments/gqlTypes/BaseCollection";
import { WithList } from "../types";
import { CollectionList } from "./CollectionList";

export class CollectionsAPI
  implements
    WithList<CollectionListQuery, BaseCollection, CollectionListVariables> {
  client: ApolloClient<any>;

  constructor(client: ApolloClient<any>) {
    this.client = client;
  }

  /**
   * Method returning list of collections with ability to request next page
   * @param params List parameters
   */
  getList = (variables: CollectionListVariables): CollectionList => {
    const list = new CollectionList(this.client);

    list.init(variables);

    return list;
  };
}
