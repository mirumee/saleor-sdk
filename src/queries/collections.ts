import gql from "graphql-tag";
import { baseCollectionFragment } from "../fragments/collections";
import { pageInfo } from "../fragments/pageInfo";

export const collections = gql`
  ${baseCollectionFragment}
  ${pageInfo}
  query CollectionList($first: Int, $after: String) {
    collections(first: $first, after: $after) {
      edges {
        node {
          ...BaseCollection
        }
      }
      pageInfo {
        ...PageInfo
      }
    }
  }
`;
