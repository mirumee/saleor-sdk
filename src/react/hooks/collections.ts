import { CollectionList } from "../../api/collections/CollectionList";
import { makeList } from "./utils";

export const useCollectionList = makeList(client => new CollectionList(client));
