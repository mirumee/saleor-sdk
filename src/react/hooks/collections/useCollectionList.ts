import { useApolloClient } from "react-apollo";
import { useEffect, useRef, useState } from "react";
import ApolloClient from "apollo-client";
import {
  CollectionList as CollectionListQuery,
  CollectionListVariables,
} from "../../../queries/gqlTypes/CollectionList";
import { BaseCollection } from "../../../fragments/gqlTypes/BaseCollection";
import { ListParameters } from "../../../api/types";
import { collections } from "../../../queries/collections";
import { CollectionList } from "../../../api/collections/CollectionList";

function query(client: ApolloClient<any>) {
  return (variables: CollectionListVariables) =>
    client.query<CollectionListQuery, CollectionListVariables>({
      query: collections,
      variables,
    });
}

function useCollectionList(params: ListParameters) {
  const client = useApolloClient();
  const [data, setData] = useState<BaseCollection[] | undefined>(undefined);
  const list = useRef<CollectionList>();
  const willMount = useRef(true);

  async function initList() {
    list.current = new CollectionList(query(client), params.count);
    await list.current.init({
      after: params.endCursor,
      first: params.count,
    });
    setData(list.current.data);
  }

  useEffect(() => {
    if (!willMount.current) {
      initList();
    }
  }, [params.count, params.endCursor]);

  if (willMount.current) {
    initList();
    willMount.current = false;
  }

  const next = async () => {
    const nextData = await list.current?.next();
    setData(nextData);
  };

  return {
    current: list.current?.current,
    data,
    loading: list.current?.loading,
    next,
    pageInfo: list.current?.pageInfo,
  };
}

export default useCollectionList;
