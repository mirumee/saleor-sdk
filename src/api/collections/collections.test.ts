import setupAPI from "../../../testUtils/api";
import { CollectionsAPI } from "./collections";

const { client } = setupAPI();

describe("Collection object", () => {
  it("can get a list of collections", async () => {
    const collectionsAPI = new CollectionsAPI(client);
    const list = collectionsAPI.getList({
      count: 20,
    });

    expect(list.data).toBeUndefined();
    expect(list.loading).toBe(true);
    await list.current;

    expect(list.data).toMatchSnapshot();
    expect(list.loading).toBe(false);
  });

  it("can get new page", async () => {
    const collectionsAPI = new CollectionsAPI(client);
    const list = collectionsAPI.getList({
      count: 1,
    });

    expect(list.data).toBeUndefined();
    expect(list.loading).toBe(true);
    await list.current;

    expect(list.data).toMatchSnapshot();
    expect(list.loading).toBe(false);

    list.next();

    expect(list.loading).toBe(true);

    await list.current;

    expect(list.data).toMatchSnapshot();
    expect(list.loading).toBe(false);
  });
});
