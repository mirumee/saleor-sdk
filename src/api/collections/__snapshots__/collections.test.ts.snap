// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection object can get a list of collections 1`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjoy",
    "name": "Winter sale",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "winter-sale",
  },
]
`;

exports[`Collection object can get new page 1`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
]
`;

exports[`Collection object can get new page 2`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjoy",
    "name": "Winter sale",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "winter-sale",
  },
]
`;
